src/qcrypto/simbasics.py

```diff
@@ -1,132 +1,424 @@
-                If not given, will default to 0 and pi/2.
+            init_method (str): Initialziation method
+
+        Returns:
+            None
+
         """
-        self.qubits = []
-        self.measurements = []
-        self.key = []
+        if self.init_method not in ["zeros", "random"]:
+            raise ValueError("Invalid initialization method.")
+        if self.init_method == "zeros":
+            self._state = np.zeros(2**self.num_qubits, dtype=np.complex_)
+            self._state[0] = 1
+        elif self.init_method == "random":
+            self._state = np.random.rand(2**self.num_qubits) + 1j * np.random.rand(
+                2**self.num_qubits
+            )
+            self._normalize_state()
 
-        if numqubits is not None or numqubits != 0:
-            if basis_selection is None:
-                self.basis_selection = [0, np.pi / 4]
-            else:
-                self.basis_selection = basis_selection
+    def measure(self, qubit_idx: int):
+        """
+        Measure ths qubit_idx'th qubit, calculating the probability of and, with these, returning 1 or 0.
+
+        Args:
+            qubit_idx (int): Index identifying the qubit to be measured
 
-            qubitbases = np.random.choice(self.basis_selection, numqubits)
+        Returns:
+            Outcome of the measurement, either 0 or 1
+        """
 
-            for basis in qubitbases:
-                # theta = 0 -> qubit = 0 state
-                # theta = pi -> qubit = 1 state
-                qubit = Qubit(np.random.choice([0, np.pi]), basis, phi=0)
-                self.qubits.append(qubit)
+        probs_0, probs_1 = self._calculate_measurement_probs(qubit_idx)
+        outcome = np.random.choice([0, 1], p=[probs_0, probs_1])
+        self._update_state_post_measurement(qubit_idx, outcome)
+        return outcome
 
-    def project(self, qubit, base):
+    def measure_all(self, order):
         """
-        Projects the given qubit unto the agent's measuring basis
+        Measures all of the qubits
+
+        Args:
+            order (str): Specifies the order in which the qubits will be measured.
+                "simult" = all qubits measured simultaneously
+                "sequential" = qubits measured in sequential order (first 0, second 1, etc.)
+
+        Returns:
+            Outcome of the measurements done. Array of 0's and 1's equal in length to the number of qubits in the system.
         """
-        newqubit = qubit
-        base_delta = base - newqubit.base
+        if order == "simult":
+            outcome = np.random.choice(
+                np.arange(len(self._state)), p=self._calculate_measurement_probs()
+            )
+            self._state.fill(0 + 0j)
+            self._state[outcome] = 1 + 0j
+            outcome_arr = np.array(
+                list(
+                    "0" * (self.num_qubits - len(bin(outcome)[2:])) + bin(outcome)[2:]
+                ),
+                dtype=int,
```

Comment:
- `outcome` is an integer who's binary representation, given by `bin(outcome)` has a "prefix" of `0b`, hence the slicing in `bin(outcome)[2:].
-  `"0" * (self.num_qubits - len(bin(outcome)[2:]))` pads the binary string with zeros on the left to make sure it has length equal to `self.num_qubits`, which is necessary because the binary string might be shorter than this, leading to zeros not included in the binary representation.
- `+ bin(outcome)[2:]` just adds the outcome binary string to this.
- This is all converted to an array so that the final result is an array where each element is a 1 or 0, representing the measurement result of each of the qubits.

Reply:
I'm sure there's easier ways of doing this, but I think it was late at night at night when I wrote this. Open to suggestions.